pageserver/src/layered_repository.rs

```diff
@@ -1708,7 +1708,11 @@ impl LayeredTimeline {
             }
 
             // 3. Is there a later on-disk layer for this relation?
-            if !l.is_dropped() && !layers.newer_image_layer_exists(l.get_seg_tag(), l.get_end_lsn())
+            if !l.is_dropped()
+                && !layers.newer_image_layer_exists(
+                    l.get_seg_tag(),
+                    Lsn::max(l.get_end_lsn(), disk_consistent_lsn),
```

Comment:
This doesn't look right to me either.

Please write a test case to demonstrate the problem, that'll be really helpful now to verify the fix, and to make sure we don't reintroduce the same issue later.

Reply:
Frankly speaking it is not so easy: it is necessary that some checkpointer thread create some new persistent layers but not yet advance disk consistent lsn. I "easily" reproduce it with 1Tb relation and out-of-disk space error. But I don't know how to reproduce it with small relation without adding some special hooks in repository code. 

It will be also useful to add tests for out-of-space errors handling. It can be done by running tests on tmpfs partition with limited size. Such environment can also be used to test this problem with GC. But I am not sure that we should do it right now.